### Random Forest

#### Load the dataset and packages

In [54]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score, confusion_matrix

df = pd.read_csv('data-drugs.csv')
df.head()

Age Sex      BP Cholesterol  Na_to_K   Drug
0   23   F    HIGH        HIGH   25.355  drugY
1   47   M     LOW        HIGH   13.093  drugC
2   47   M     LOW        HIGH   10.114  drugC
3   28   F  NORMAL        HIGH    7.798  drugX
4   61   F     LOW        HIGH   18.043  drugY

In [55]:
df.isnull().any()
#df.isnull().values.any()

Age            False
Sex            False
BP             False
Cholesterol    False
Na_to_K        False
Drug           False
dtype: bool

In [56]:
df.isnull().values.sum()

0

In [57]:
#Showing the null values as per attributes
null_columns=df.columns[df.isnull().any()]
df[null_columns].isnull().sum()

Series([], dtype: float64)

In [58]:
df['BP'] = df['BP'].map({'HIGH':2, 'LOW':0, 'NORMAL':1})
df['Sex'] = df['Sex'].map({'M':1, 'F':0})
df['Cholesterol'] = df['Cholesterol'].map({'HIGH':1, 'NORMAL':0})

In [63]:
#X = df[['Age','Sex', 'BP', 'Cholesterol', 'Na_to_K']]
X = df.loc[:, df.columns != 'Drug']
y =df['Drug'] 
X

Age  Sex  BP  Cholesterol  Na_to_K
0     23    0   2            1   25.355
1     47    1   0            1   13.093
2     47    1   0            1   10.114
3     28    0   1            1    7.798
4     61    0   0            1   18.043
..   ...  ...  ..          ...      ...
195   56    0   0            1   11.567
196   16    1   0            1   12.006
197   52    1   1            1    9.894
198   23    1   1            0   14.020
199   40    0   0            0   11.349

[200 rows x 5 columns]

In [64]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=3)
rf = RandomForestClassifier(n_estimators=10, max_depth=5, random_state=0)
rf.fit(X_train, y_train)

RandomForestClassifier(max_depth=5, n_estimators=10, random_state=0)

In [66]:
y_pred = rf.predict(X_test)
y_pred

array(['drugY', 'drugX', 'drugX', 'drugX', 'drugX', 'drugC', 'drugY',
       'drugA', 'drugB', 'drugA', 'drugY', 'drugA', 'drugY', 'drugY',
       'drugX', 'drugY', 'drugX', 'drugX', 'drugB', 'drugX', 'drugX',
       'drugY', 'drugY', 'drugY', 'drugX', 'drugB', 'drugY', 'drugY',
       'drugA', 'drugX', 'drugB', 'drugC', 'drugC', 'drugX', 'drugX',
       'drugC', 'drugY', 'drugX', 'drugX', 'drugX', 'drugA', 'drugY',
       'drugC', 'drugY', 'drugA', 'drugY', 'drugY', 'drugY', 'drugY',
       'drugY', 'drugB', 'drugX', 'drugY', 'drugX', 'drugY', 'drugY',
       'drugA', 'drugX', 'drugY', 'drugX'], dtype=object)

In [73]:
print('Classification Report :\n', classification_report(y_test, y_pred))
print('\nAccuracy Score:{0:.4f}'.format(accuracy_score(y_test,y_pred)))
print('Confusion Matrix:\n', confusion_matrix(y_test, y_pred))

Classification Report :
               precision    recall  f1-score   support

       drugA       1.00      1.00      1.00         7
       drugB       1.00      1.00      1.00         5
       drugC       1.00      1.00      1.00         5
       drugX       1.00      0.95      0.98        21
       drugY       0.96      1.00      0.98        22

    accuracy                           0.98        60
   macro avg       0.99      0.99      0.99        60
weighted avg       0.98      0.98      0.98        60


Accuracy Score:0.9833
Confusion Matrix:
 [[ 7  0  0  0  0]
 [ 0  5  0  0  0]
 [ 0  0  5  0  0]
 [ 0  0  0 20  1]
 [ 0  0  0  0 22]]


#### Finally, our random forest accuracy was found to be 98%